File: Assets/Scripts/LoadOBJ/RuntimeObjImporter.cs
using UnityEngine;
using System.Collections.Generic;
using System.Globalization;
using System;

public class LoadedObjData
{
    public Mesh mesh;
    public string mtlFileName;
}

public static class RuntimeObjImporter
{
    public static LoadedObjData Import(string objContent, bool flipUV = true, bool flipFaces = false)
    {
        List<Vector3> vertices = new List<Vector3>();
        List<Vector3> normals = new List<Vector3>();
        List<Vector2> uv = new List<Vector2>();
        
        List<Vector3> newVertices = new List<Vector3>();
        List<Vector3> newNormals = new List<Vector3>();
        List<Vector2> newUVs = new List<Vector2>();
        List<int> triangles = new List<int>();

        string mtlFileName = "";

        string[] lines = objContent.Split(new char[] { '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);

        foreach (string line in lines)
        {
            string l = line.Trim();
            if (string.IsNullOrEmpty(l) || l.StartsWith("#"))
                continue;

            string[] parts = l.Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
            if (parts.Length == 0) continue;

            string type = parts[0];

            if (type == "mtllib")
            {
                // Material library file
                if (parts.Length > 1)
                    mtlFileName = parts[1];
            }
            else if (type == "v")
            {
                float x = float.Parse(parts[1], CultureInfo.InvariantCulture);
                float y = float.Parse(parts[2], CultureInfo.InvariantCulture);
                float z = float.Parse(parts[3], CultureInfo.InvariantCulture);
                vertices.Add(new Vector3(-x, y, z));
            }
            else if (type == "vn")
            {
                float x = float.Parse(parts[1], CultureInfo.InvariantCulture);
                float y = float.Parse(parts[2], CultureInfo.InvariantCulture);
                float z = float.Parse(parts[3], CultureInfo.InvariantCulture);
                normals.Add(new Vector3(-x, y, z));
            }
            else if (type == "vt")
            {
                float u = float.Parse(parts[1], CultureInfo.InvariantCulture);
                float v = float.Parse(parts[2], CultureInfo.InvariantCulture);
                
                if (flipUV)
                    uv.Add(new Vector2(u, 1f - v));
                else
                    uv.Add(new Vector2(u, v));
            }
            else if (type == "f")
            {
                int[] vIndices = new int[parts.Length - 1];
                int[] tIndices = new int[parts.Length - 1];
                int[] nIndices = new int[parts.Length - 1];

                for (int i = 0; i < parts.Length - 1; i++)
                {
                    string[] indices = parts[i + 1].Split('/');
                    
                    if (indices.Length >= 1 && !string.IsNullOrEmpty(indices[0]))
                        vIndices[i] = int.Parse(indices[0]);

                    if (indices.Length >= 2 && !string.IsNullOrEmpty(indices[1]))
                        tIndices[i] = int.Parse(indices[1]);

                    if (indices.Length >= 3 && !string.IsNullOrEmpty(indices[2]))
                        nIndices[i] = int.Parse(indices[2]);
                }
                
                for (int i = 0; i < vIndices.Length - 2; i++)
                {
                    if (flipFaces)
                    {
                        // Reverse winding order: 0, 2, 1
                        AddVertexData(vIndices[0], tIndices[0], nIndices[0], vertices, uv, normals, newVertices, newUVs, newNormals, triangles);
                        AddVertexData(vIndices[i + 2], tIndices[i + 2], nIndices[i + 2], vertices, uv, normals, newVertices, newUVs, newNormals, triangles);
                        AddVertexData(vIndices[i + 1], tIndices[i + 1], nIndices[i + 1], vertices, uv, normals, newVertices, newUVs, newNormals, triangles);
                    }
                    else
                    {
                        // Standard winding order: 0, 1, 2
                        AddVertexData(vIndices[0], tIndices[0], nIndices[0], vertices, uv, normals, newVertices, newUVs, newNormals, triangles);
                        AddVertexData(vIndices[i + 1], tIndices[i + 1], nIndices[i + 1], vertices, uv, normals, newVertices, newUVs, newNormals, triangles);
                        AddVertexData(vIndices[i + 2], tIndices[i + 2], nIndices[i + 2], vertices, uv, normals, newVertices, newUVs, newNormals, triangles);
                    }
                }
            }
        }

        Mesh mesh = new Mesh();
        if (newVertices.Count > 65000)
            mesh.indexFormat = UnityEngine.Rendering.IndexFormat.UInt32;
            
        mesh.vertices = newVertices.ToArray();
        
        if (newUVs.Count > 0)
            mesh.uv = newUVs.ToArray();
            
        if (newNormals.Count > 0)
            mesh.normals = newNormals.ToArray();
            
        mesh.triangles = triangles.ToArray();
        
        if (newNormals.Count == 0)
            mesh.RecalculateNormals();
            
        mesh.RecalculateBounds();

        return new LoadedObjData { mesh = mesh, mtlFileName = mtlFileName };
    }

    private static int ResolveObjIndex(int idx, int count)
    {
        if (idx > 0) return idx;
        if (idx < 0) return count + idx + 1;
        return 0;
    }

    private static void AddVertexData(
        int vIdx, int tIdx, int nIdx,
        List<Vector3> sourceV, List<Vector2> sourceT, List<Vector3> sourceN,
        List<Vector3> destV, List<Vector2> destT, List<Vector3> destN,
        List<int> destTriangles)
    {
        vIdx = ResolveObjIndex(vIdx, sourceV.Count);
        tIdx = ResolveObjIndex(tIdx, sourceT.Count);
        nIdx = ResolveObjIndex(nIdx, sourceN.Count);

        if (vIdx > 0 && vIdx <= sourceV.Count)
            destV.Add(sourceV[vIdx - 1]);
        else
            destV.Add(Vector3.zero);

        if (tIdx > 0 && tIdx <= sourceT.Count)
            destT.Add(sourceT[tIdx - 1]);
        else if (destT.Count > 0 || sourceT.Count > 0)
            destT.Add(Vector2.zero);

        if (nIdx > 0 && nIdx <= sourceN.Count)
            destN.Add(sourceN[nIdx - 1]);
        else if (destN.Count > 0 || sourceN.Count > 0)
            destN.Add(Vector3.up);

        destTriangles.Add(destV.Count - 1);
    }
}

File: Assets/Lua/RuntimeObjImporter.lua
local RuntimeObjImporter = {}

-- 获取 Unity 类型引用
local CS = CS or {}
local UnityEngine = CS.UnityEngine
local Vector3 = UnityEngine and UnityEngine.Vector3
local Vector2 = UnityEngine and UnityEngine.Vector2
local Mesh = UnityEngine and UnityEngine.Mesh
local IndexFormat = UnityEngine and UnityEngine.Rendering and UnityEngine.Rendering.IndexFormat

-- 如果不在 Unity 环境中，提供简单的 Mock 实现（防止报错，方便测试）
if not UnityEngine then
    Vector3 = { 
        zero = {x=0, y=0, z=0}, 
        up = {x=0, y=1, z=0},
        new = function(x,y,z) return {x=x, y=y, z=z} end 
    }
    setmetatable(Vector3, { __call = function(_, x,y,z) return Vector3.new(x,y,z) end })
    
    Vector2 = { 
        zero = {x=0, y=0},
        new = function(x,y) return {x=x, y=y} end 
    }
    setmetatable(Vector2, { __call = function(_, x,y) return Vector2.new(x,y) end })
    
    Mesh = { new = function() 
        return { 
            RecalculateNormals = function() end,
            RecalculateBounds = function() end
        } 
    end }
    IndexFormat = { UInt32 = 1 }
end

-- 字符串分割工具函数
local function SplitString(str, separator)
    local result = {}
    -- 使用 pattern 匹配分割
    -- 注意：特殊字符需要转义
    local pattern = "(.-)" .. separator
    for match in (str .. separator):gmatch(pattern) do
        table.insert(result, match)
    end
    return result
end

local function SplitByWhitespace(str)
    local result = {}
    for match in str:gmatch("%S+") do
        table.insert(result, match)
    end
    return result
end

-- 添加顶点数据辅助函数
local function AddVertexData(vIdx, tIdx, nIdx, sourceV, sourceT, sourceN, destV, destT, destN, destTriangles)
    -- Lua 数组从 1 开始，OBJ 索引从 1 开始
    -- sourceV[vIdx]
    if vIdx > 0 and vIdx <= #sourceV then
        table.insert(destV, sourceV[vIdx])
    else
        table.insert(destV, Vector3.zero)
    end

    if tIdx > 0 and tIdx <= #sourceT then
        table.insert(destT, sourceT[tIdx])
    elseif #destT > 0 or #sourceT > 0 then
        -- 如果有任何纹理坐标，保持数组对齐
        table.insert(destT, Vector2.zero)
    end

    if nIdx > 0 and nIdx <= #sourceN then
        table.insert(destN, sourceN[nIdx])
    elseif #destN > 0 or #sourceN > 0 then
        table.insert(destN, Vector3.up)
    end

    -- destV 的长度就是当前新顶点的 1-based 索引
    -- Unity Mesh.triangles 期望的是 0-based 索引
    table.insert(destTriangles, #destV - 1)
end

-- 主导入函数
function RuntimeObjImporter.Import(objContent)
    local vertices = {}
    local normals = {}
    local uv = {}
    
    local newVertices = {}
    local newNormals = {}
    local newUVs = {}
    local triangles = {}
    
    local mtlFileName = ""
    
    -- 按行处理
    for line in objContent:gmatch("[^\r\n]+") do
        local l = line:gsub("^%s*(.-)%s*$", "%1") -- Trim
        
        -- 跳过空行和注释
        if l ~= "" and not l:find("^#") then
            local parts = SplitByWhitespace(l)
            if #parts > 0 then
                local type = parts[1]
                
                if type == "mtllib" then
                    if #parts > 1 then
                        mtlFileName = parts[2]
                    end
                elseif type == "v" then
                    local x = tonumber(parts[2]) or 0
                    local y = tonumber(parts[3]) or 0
                    local z = tonumber(parts[4]) or 0
                    -- 坐标转换: new Vector3(-x, y, z)
                    table.insert(vertices, Vector3(-x, y, z))
                elseif type == "vn" then
                    local x = tonumber(parts[2]) or 0
                    local y = tonumber(parts[3]) or 0
                    local z = tonumber(parts[4]) or 0
                    -- 坐标转换: new Vector3(-x, y, z)
                    table.insert(normals, Vector3(-x, y, z))
                elseif type == "vt" then
                    local u = tonumber(parts[2]) or 0
                    local v = tonumber(parts[3]) or 0
                    table.insert(uv, Vector2(u, v))
                elseif type == "f" then
                    local vIndices = {}
                    local tIndices = {}
                    local nIndices = {}
                    
                    -- parts[1] 是 "f"，后续是索引数据
                    for i = 2, #parts do
                        local indexStr = parts[i]
                        local subParts = SplitString(indexStr, "/")
                        
                        -- v 索引
                        if #subParts >= 1 and subParts[1] ~= "" then
                            table.insert(vIndices, tonumber(subParts[1]))
                        else
                            -- OBJ 索引如果不合法通常设为0或者忽略，这里用0占位
                            table.insert(vIndices, 0)
                        end
                        
                        -- vt 索引
                        if #subParts >= 2 and subParts[2] ~= "" then
                            table.insert(tIndices, tonumber(subParts[2]))
                        else
                            table.insert(tIndices, 0)
                        end
                        
                        -- vn 索引
                        if #subParts >= 3 and subParts[3] ~= "" then
                            table.insert(nIndices, tonumber(subParts[3]))
                        else
                            table.insert(nIndices, 0)
                        end
                    end
                    
                    -- 三角剖分 (Triangulate)
                    -- 假设是凸多边形，使用扇形剖分
                    for i = 1, #vIndices - 2 do
                        AddVertexData(vIndices[1], tIndices[1], nIndices[1], vertices, uv, normals, newVertices, newUVs, newNormals, triangles)
                        AddVertexData(vIndices[i + 1], tIndices[i + 1], nIndices[i + 1], vertices, uv, normals, newVertices, newUVs, newNormals, triangles)
                        AddVertexData(vIndices[i + 2], tIndices[i + 2], nIndices[i + 2], vertices, uv, normals, newVertices, newUVs, newNormals, triangles)
                    end
                end
            end
        end
    end
    
    -- 创建 Mesh 对象
    local mesh = Mesh.new()
    
    -- 设置索引格式
    if #newVertices > 65000 then
        mesh.indexFormat = IndexFormat.UInt32
    end
    
    -- 赋值 Mesh 数据
    -- 注意: 具体赋值方式取决于 Lua 绑定库 (xLua, ToLua 等)
    -- 通常可以直接赋值 Lua Table 或者需要转为 C# Array/List
    -- 这里假设可以直接赋值
    mesh.vertices = newVertices
    
    if #newUVs > 0 then
        mesh.uv = newUVs
    end
    
    if #newNormals > 0 then
        mesh.normals = newNormals
    end
    
    mesh.triangles = triangles
    
    if #newNormals == 0 then
        mesh:RecalculateNormals()
    end
    
    mesh:RecalculateBounds()
    
    return { mesh = mesh, mtlFileName = mtlFileName }
end

return RuntimeObjImporter
